src/ldc/llvmasm.di

```diff
@@ -18,3 +18,58 @@ pragma(LDC_inline_asm)
         __asmtuple_t!(T) __asmtuple(const(char)[] asmcode, const(char)[] constraints, ...);
     }
 }
+
+
+/++
+ + Calls a function whose body is written in the LLVM assembly language (IR).
+ +
+ + Template params:
+ +   s = the LLVM IR function body code
+ +   R = the function return type
+ +   P... = the types of the function arguments
+ +
+ + If the return type is `void` then `ret void` is automatically added at the
+ + end of the function body by LDC.
+ +
+ + Example:
+ + ---
+ + import ldc.llvmasm;
+ + int add(int a, int b)
+ + {
+ +     return __ir!(`%r = add i32 %0, %1
+ +                   ret i32 %r`, int)(a, b);
+ + }
+ + ---
+ +/
+pragma(LDC_inline_ir)
+    R __ir(string s, R, P...)(P params) pure nothrow @nogc;
+
+/// Ditto
+pragma(LDC_inline_ir)
+    R __ir_trusted(string s, R, P...)(P params) @trusted pure nothrow @nogc;
```

Comment:
Why special treatment for trusted? User can write function which access global var or call to other D functions and violate other attributes. maybe just discourage usage of simd.inlineIR and make it private eventually so user have to write his own function with correct attributes. OR leave generic `__ir` function without any attributes so user will need to write his own version only if he can guarantee some attributes.

Reply:
The point of defining `__asm` and `__ir` publically in this module is that the user should not be needing to write `pragma(...)` and define his own inline IR (or inline asm) function.
I don't think there is a perfect solution for dealing with attributes for all cases.  We have to make decisions about the attributes here because automatically inferring attributes is very hard. Here I choose to make the same default attribute decisions as for `__asm`. We know they are wrong in some cases, but having the user pass them as arguments becomes quite tedious (and unreadable), for example:
```d
__ir!("llvm ir code", RetType, true /*trusted?*/, false /*pure?*/, false /*nothrow?*/, true /*nogc?*/)(a, b, c);
```
Defaults for the template parameters are not possible I think because of the `P...` pack at the end.
But ideas on how to make this better are appreciated :)

What would user code look like if `__ir` has no attributes, but must be called in `pure` context?

Reply:
> What would user code look like if __ir has no attributes, but must be called in pure context?

Define his own version with pragma with attributes he can guarantee.

Reply:
> Define his own version with pragma with attributes he can guarantee.

The goal of `__asm` and `__ir` is exactly to not have to do that.

If we add no attributes to `__ir`, things work "OK-ish" for safe/trusted/system:
```d
pragma(LDC_inline_ir)
    R __ir(string s, R, P...)(P params);
    
int add(int a, int b) {
     return __ir!(`%r = add i32 %0, %1
                   ret i32 %r`, int)(a, b);
}

int addsafe(int a, int b) @safe {
    auto ret = (() @trusted => __ir!(`%r = add i32 %0, %1
                                      ret i32 %r`, int)(a, b))();
    return ret;
}
```

Weka has 25 lines of complex code to be able to call functions as if they have certain attributes, leading to code like `auto ret = as!"@nogc nothrow"({ __ir(...); });` . It is very non-trivial, but 'works' but... not for `pure`.

Reply:
I agree, these declarations are for convenience (as well as documentation to some degree), and the possibly wrong extra attributes simplify things with D's attribute hell. Considering these attributes are purely cosmetics in terms of codegen (correct me if I'm wrong), they shouldn't do much harm in the most likely extremely rare usages.

Reply:
Right now, I'm thinking about making `__ir` and `__ir_pure`, both with `nothrow, @nogc, @trusted`. `pure` is the only one that can potentially lead to codegen differences (optimization) in the future.
What do you think?

Reply:
Yeah, special-casing `pure` makes probably a lot more sense than `@trusted`. The same would apply to `__asm` too... ;)